mqtt/mqtt-bridge/src/client.rs

```diff
@@ -0,0 +1,223 @@
+                Duration::from_secs(1),
+                keep_alive,
+            )
+        } else {
+            Client::from_state(
+                client_id,
+                None,
+                None,
+                TcpConnection {
+                    address: address.into(),
+                    token_source,
+                },
+                Duration::from_secs(1),
+                keep_alive,
+            )
+        };
+
+        MqttClient {
+            client,
+            event_handler,
+        }
+    }
+
+    pub fn shutdown_handle(&self) -> Result<ShutdownHandle, ShutdownError> {
+        self.client
+            .shutdown_handle()
+            .map_or(Err(ShutdownError::ClientDoesNotExist), |shutdown_handle| {
+                Ok(ShutdownHandle(shutdown_handle))
+            })
+    }
+
+    pub async fn handle_events(mut self) -> Result<(), ClientConnectError> {
+        while let Some(event) = self.client.try_next().await.unwrap_or_else(|e| {
+            error!(message = "failed to poll events", error=%e);
+            None
+        }) {
+            if let Err(e) = self.event_handler.handle_event(event).await {
+                error!("error processing event {}", e);
+            }
+        }
+
+        Ok(())
+    }
+
+    pub async fn subscribe(&mut self, topics: Vec<String>) -> Result<(), ClientConnectError> {
+        debug!("subscribing to topics");
+        let subscriptions = topics.iter().map(|topic| proto::SubscribeTo {
+            topic_filter: topic.to_string(),
+            qos: proto::QoS::AtLeastOnce,
+        });
+
+        for subscription in subscriptions {
+            debug!("subscribing to topic {}", subscription.topic_filter);
+            self.client
+                .subscribe(subscription)
+                .map_err(ClientConnectError::SubscribeFailure)?;
+        }
+
+        let mut subacks: HashSet<_> = topics.iter().collect();
+        if subacks.is_empty() {
+            debug!("no topics to subscribe to");
+            return Ok(());
+        }
+
+        while let Some(event) = self
+            .client
+            .try_next()
+            .await
+            .map_err(ClientConnectError::PollClientFailure)?
+        {
+            if let Event::SubscriptionUpdates(subscriptions) = event {
+                for subscription in subscriptions {
+                    if let SubscriptionUpdateEvent::Subscribe(sub) = subscription {
+                        subacks.remove(&sub.topic_filter);
+                    }
+                }
+
+                if subacks.is_empty() {
+                    debug!("successfully subscribed to all topics");
+                    return Ok(());
+                }
+            }
+        }
+
+        error!("failed to subscribe to topics");
+        Err(ClientConnectError::MissingSubacks(
+            subacks
+                .into_iter()
+                .map(std::string::ToString::to_string)
+                .collect(),
+        ))
+    }
+}
```

Comment:
Since this logic is being duplicated, we should have a plan to consolidate if we are not going to do it now.

Reply:
yes, the plan is to consolidate at a later time